### DISCUSSION TITLE
Dual ST7789 240x240 displays

### DISCUSSION THREAD

**Discussion by Boyeen** [Bodmer/TFT_eSPI#1546] [Compatible displays and setup files]
So,  decided to try a pair of ST7789 displays on an ESP-32. Sadly no joy using this lib. Here's what I have done:
Modified the display modules to bring out the CS pin. From some Bodmer comments I discovered each display needs dedicated RST as well as CS.  Makes sense, as the driver resets each controller after init. I know this lib can't handle multiple instances and I have had a lot of success with up to 5 ST7735 displays on a single SPI by driving the CS pins directly.  However, the ST7789 has defeated me. Each individual displays works by itself but not both on the same SPI bus. Just ordered a logic analyzer and will dive into the SPI bus I guess...  I know this display has some SPI-CLK issues and perhaps this is just not possible...

BTW My breadboarded design works great with the Adafruit ST7789 driver (two instances, one per display) but it's unbelievably slow, Bodmer has me spoiled with speed :-).  

Any suggestions?

**Comment by Bodmer** [comment]
I can't explain this. I have had 2 displays working and I see you have as well.

If the two displays are connected with a single chip select line do both work and show the same thing?

You can init/reset all displays together but subsequently the chip selects must be controlled depending on the display to be updated.

**Reply by Boyeen** [reply]
Thanks for the reply Bodmer. Before I forget, what a fantastic library. Superb work...
To answer your question, yes, I can init and operate both displays in parallel but only if I let the library manage CS.
Manual CS does not work, one display or  both in parallel.  

Tearing my hair out I resorted to trying to duplicate the CS init code in the library... and I came across your comment 
"Some displays need a high toggle of the CS".    Well, what better idea do I have?  so I started from scratch with the TFT printing
 sketch and added a toggle of CS just before init().   What do you know, it worked...  I can init() both displays in parallel or alone.
I can even run the sketch on alternate displays. Of course it's a trivial sketch.

Back in my main sketch it ALSO works for trivial activity like cycling thru different screen colors but anything more complex
like loading jpeg images seems to crash the display, single or parallel.  One step forward...

**Reply by Bodmer** [reply]
If the jpeg is done by DMA then you must wait for DMA to finish before changing the chip select. You may also need a power supply that can power both displays and the processor. Sounds like you have made a lot of progress.

**Reply by Boyeen** [reply]
Thanks, do you have any more info on the "toggle" requirements? Is it just for init() or EVERY time I assert CS? Also does it apply to deassert too?

Thanks

**Reply by husnoo** [reply]
I'm going to try to drive two displays as well, and mine don't have a CS pin. I've managed to wire up both displays in parallel by sharing all the pins. I noticed if I pull out the MOSI on each individually, I can stop it from being updated... so I'm thinking of using pair of transistors to switch on/off the MOSI using a GPIO. Dunno if it'll work, I'll let you know!

edit: oh sorry Google led me here, but I'm using the adafruit library.

**Reply by Bodmer** [reply]
Switching MOSI will not work, switch SCLK instead.

**Comment by bcourt** [comment]
I am have the same problem Has this been sorted ??

I am running 4 displays (ST7789) from ESP32 with TFT_eSPI and getting the same problem. i can write to each display by using a CS for each display eg CS1,CS2,CS3,CS4.
all works if init then write to the display one at a time. the problem is the display goes blank when i move to the next display.
